File: web/crossplatform/RXCypher/RXCypher.js
/*
 * Title:		RXCypher
 * Description:	Simple and fast stream cipher algorithm - JavaScript implementation
 * Author:		Jonathan Voss
 * Date:		8/7/2012
 * Namespace:	github.com/k98kurz
 * License:		MIT
 *
 * Basic byte-level algorithm
 * t = plaintext byte, c = ciphertext byte, k = encryption key byte
 *
 *		function enc (t, k) { return (t+k>255 ? t+k-255 : t+k)^k; }
 *		function dec (c, k) { return (c^k)-k<0 ? (c^k)-k+255 : (c^k)-k; }
 *
 *
 * Full singleton
 * Methods:
 *		enc ( sString, sKey )
 *			encrypts plaintext sString with encryption key sKey
 *		dec ( sCypherText, sKey )
 *			decrypts cipher text sCypherText with decryption key sKey
*/

var RXCypher = {
	encrypt: function(plaintext, key) {
		var ciphertext = "", i, j, g, t, k;
		for (i=0, j=0; i<plaintext.length; i++, j++) {
			j = (j==key.length) ? 0 : j;
			t = plaintext.charCodeAt(i);
			k = key.charCodeAt(j);
			g = (t+k>255 ? t+k-255 : t+k)^k;
			ciphertext += g.toString(16);
		}
		return ciphertext;
	},
	decrypt: function(ciphertext, key) {
		var plaintext = "", i, j, g, c, k;
		for (i=0, j=0; i<ciphertext.length; i++, j++) {
			j = (j==key.length) ? 0 : j;
			c = parseInt(ciphertext.charAt(i)+ciphertext.charAt(++i),16);
			k = key.charCodeAt(j);
			g = (c^k)-k<0 ? (c^k)-k+255 : (c^k)-k;
			plaintext += String.fromCharCode(g);
		}
		return plaintext;
	}
};

File: java/RXCypher/RXCypher.java
package com.noobgrinder.jonathan.rxcypher;

public class RXCypher {
	private String key;
	
	RXCypher (String k) throws Exception {
		if (k.length()>0) {
			key = k;
		} else {
			throw new Exception("Must supply an encryption key");
		}
	}
	
	public String encrypt (String str) {
		String result = "";
		int i; int k;
		for (i=0, k=0; i<str.length(); i++, k++) {
			if (k>=key.length()) { k =0; }
			result += enc(str.charAt(i), key.charAt(k));
		}
		return result;
	}
	
	public String decrypt (String str) {
		String result = "";
		int i; int k;
		for (i=0, k=0; i<str.length(); i++, k++) {
			if (k>=key.length()) { k = 0; }
			result += dec(str.charAt(i), key.charAt(k));
		}
		return result;
	}
	
	private char enc (char t, char k) {
		int c;
		c = ((int)k+(int)t>255) ? k+t-255 : k+t;
		c = c^(int)k;
		return (char) c;
	}
	private char dec (char t, char k) {
		int c;
		c = (int)k^(int)t;
		return (char) ((c-(int)k<0) ? c-(int)k+255 : c-(int)k);
	}
}
